src/python/pants/backend/cc/target_types.py

```diff
@@ -76,9 +151,93 @@ class CCSourcesGeneratorTarget(TargetFilesGenerator):
     )
     generated_target_cls = CCSourceTarget
     copied_fields = COMMON_TARGET_FIELDS
-    moved_fields = (Dependencies,)
+    moved_fields = (
+        CCCompileFlagsField,
+        CCDefinesField,
+        CCDependenciesField,
+        CCLanguageField,
+    )
     help = "Generate a `cc_source` target for each file in the `sources` field."
 
 
+# -----------------------------------------------------------------------------------------------
+# `cc_library` and `cc_binary` targets
+# -----------------------------------------------------------------------------------------------
+
+
+class CCLinkTypeField(StringField):
+    alias = "link_type"
+    default = "shared"
+    valid_choices = ("shared", "static")
+    help = softwrap(
+        """
+        This field determines the link basis for `cc_library` targets.
+        """
+    )
+
+
+class CCLinkFlagsField(StringSequenceField):
+    alias = "link_flags"
+    help = softwrap(
+        """
+        TODO
+        """
+    )
+
+
+class CCHeadersField(Dependencies):
+    alias = "headers"
+    help = softwrap(
+        """
+        This field specifies public headers which are made available to dependent targets.
+        The headers should be specified as `cc_source` or `cc_sources` targets.
+        """
+    )
+
+
+class CCLibraryTarget(Target):
+    alias = "cc_library"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        CCDependenciesField,
+        CCHeadersField,
+        # CCLinkFlagsField, # TODO
+        CCLinkTypeField,
+        OutputPathField,
+    )
+    help = softwrap(
+        """
+        A static or shared cc library (depending on `link_type`).
+        """
+    )
+
+
+# TODO: Need this so building a binary doesn't build a library too
+class CCContrivedField(BoolField):
```

Comment:
I don't think this is needed anymore, after I started expanding on `library` vs `binary` targets.

Reply:
Note: I know this isn't any kind of solution, anyways, just had it in there so I could move past weird `package` redundancy while developing.